components/hab/src/main.rs

```diff
@@ -1572,6 +1603,68 @@ fn install_sources_from_matches(matches: &ArgMatches<'_>) -> Result<Vec<InstallS
         .collect()
 }
 
+fn download_idents_from_matches(matches: &ArgMatches<'_>) -> Result<Vec<PackageIdent>> {
+    match matches.values_of("PKG_IDENT") {
+        Some(ident_strings) => {
+            ident_strings.map(|t| ident_from_str_helper(t, None))
+                         .collect()
+        }
+        _ => Ok(Vec::new()), // It's not an error to have no idents on command line
+    }
+}
+
+fn download_idents_from_file_matches(matches: &ArgMatches<'_>) -> Result<Vec<PackageIdent>> {
+    let mut sources: Vec<PackageIdent> = Vec::new();
+
+    if let Some(files) = matches.values_of("PKG_IDENT_FILE") {
+        for filename in files {
+            let file = match File::open(filename) {
+                Ok(f) => f,
+                Err(_e) => {
+                    return Err(Error::ArgumentError(format!("Can't open file {}", filename)))
+                }
+            };
+            let buf_reader = BufReader::new(file);
+            let packages_or_nothing: Result<Vec<Option<PackageIdent>>> =
+                buf_reader.lines()
+                          .map(|x| expand_line(&x.unwrap(), filename))
+                          .collect(); // Collect takes Vec<Result<x>> to Result<Vec<x>>
+            packages_or_nothing?.into_iter()
+                                .filter(std::option::Option::is_some)
+                                .for_each(|y| sources.push(y.unwrap()));
+        }
+    }
+    Ok(sources)
+}
```

Comment:
With the up-front validation I alluded to before, this can be simplified to basically
```rust
    let s = std::fs::read_to_string(path).map_err(habitat_core::error::Error::IO)?;

    s.lines()
     .map(|line| PackageIdent::from_str(line.trim()))
     .collect()
```
eliminating the need for `ident_from_str_helper` and `expand_line`. I'd suggest putting that logic in a shared place (probably `habitat_common`) so that it can be shared by the validation code and the parsing args into `PackageIdent`s code.